doc/monitoring_high_mem_containers/monitoring_containers.md

```diff
@@ -0,0 +1,282 @@
+3. `CONFIG_DB` can be configured to retrieve and set memory threshold of each container.
+4. Users can access the alerting status of each container via the CLI utility
+    1. Users can retrieve alerting status of a container.
+    2. Users can configure alerting status of a container.
+5. Users can access the memory threshold of each container via the CLI utility
+    1. Users can retrieve memory threshold of a container.
+    2. Users can configure memory threshold of a container.
+
+### 1.2.2 Configuration and Management Requirements
+The default alerting status and memory threshold of containers should be initialized 
+in the `init_cfg.json.j2` file. Configuration of this feature can be changed via:
+1. config_db.json
+2. CLI
+
+### 1.2.3 Fast-Reboot/Warm-Reboot Requirements
+This feature will not affect the Fast-Reboot/Warm-Reboot procedures.
+
+## 1.3 Design
+
+### 1.3.1 Basic Approach
+Monitoring the running status of critical processes and resource usage of
+docker containers depend on the Monit system tool. Since Monit natively
+provides the mechanism to check whether a process is running or not, it will
+be straightforward to monitor the critical processes in SONiC. 
+
+However, Monit only provides a mechanism to monitor the resource usage in a
+per-process level not a per-container level. As such, monitoring the resource
+usage of a docker container is not as straightforward as monitoring processes.
+In this design, we propose to utilize the mechanism with which Monit can spawn
+a process and check its return value. We will have Monit
+launch a script which reads the resource usage of a container and compares
+the resource usage with a configured threshold value of that container.
+If the current resource usage is less than the configured threshold value,
+the script will return 0 and Monit will not take action. However, if the
+resource usage exceeds the threshold, the script will return a non-zero value
+and Monit will log an alert message into the syslog.
+
+Specifically, for monitoring memory usage of a container, the workflow can be
+described as the following steps:
+1.  Monit spawns a process to execute the script `memory_checker` every 1 minute
+2.  `memory_checker` accepts <container_name> as a parameter and retrieves 
+    alerting status and memory threshold of this container from `CONFIG_DB`.
+3.  If high memory alerting is enabled and the container is running, `memory_checker` 
+    retrieves its runtime memory usage from command output of `docker stats`; 
+    Otherwiese, `memory_checker` exits and logs an message indicating
+    either alerting is disabled or specified container is not running.
+4.  If runtime memory usage is larger than memory threshold, then `memory_checker`
+    exits with non-zero value; Otherwise, `memory_checker` exits with zero value.
+5.  Monit will write an alerting message into syslog if it receives non-zero
```

Comment:
>syslog if it receives non-zero [](http://example.com/codeflow?start=46&length=30)

Need to write both zero and non-zero value to syslog. Zero is INFO, non-zero is ERROR or CRITICAL. This way we will get confidence that the monitor process is running fine.